### ISSUE TITLE
NodeJS Version

### ISSUE THREAD

**Issue by Alexander9673** [GrowtopiaNoobs/GrowtopiaServer#1253]
I was planning to make a growtopia private server based on NodeJS, there is a package i found that uses enet, but the problem is i don't know where to start. If you guys could help, just drop your Discord Tag (e.g User#0000) and i'll add you and we'll talk on how.

I am a professional programmer at nodejs, and i know what i'm doing on it.
Here is the package i found for enet: https://github.com/Memeo/enet-js

**Comment by xXAbood** [comment]
its possible but its not recommended, quite alot of packets arent fully reversed yet.
+ it would lag with 10+ players probably on different worlds

**Comment by Alexander9673** [comment]
question, why would it lag?

**Comment by xXAbood** [comment]
> question, why would it lag?

basically its because the packets wouldnt really work properly

**Comment by kuteshiki** [comment]
@Alexander9673 I'm a node.js / web dev. I'm quite good at node.js and have a discord bot with over 60k users + a full terminal emulator wroten in HTML + js!  (invite.gg/rikka, https://wearetermi.github.io/termi)
So, we both could try to copy over the code from GNGS and attempt to do this in node.js.
@xXAbood has told me about this converastion.
BUT BUT BUT BUT BUT BUT BUT BUT BUT BUT - enet-js is over NINE YEARS old at this point. It's not maintained and it would be very hard to make it in a dead npm libary.

**Comment by kuteshiki** [comment]
We could have a bigger chance with:
https://github.com/qianqians/js_enet
https://github.com/mmorrison24/enet-excursion
as they're not 9 years old.

**Comment by Alexander9673** [comment]
> @Alexander9673 I'm a node.js / web dev. I'm quite good at node.js and have a discord bot with over 60k (peak) users + a full terminal emulator wroten in HTML + js! (invite.gg/rikka, https://wearetermi.github.io/termi)
> So, we both could try to copy over the code from GNGS and attempt to do this in node.js.
> @xXAbood has told me about this converastion.
> BUT BUT BUT BUT BUT BUT BUT BUT BUT BUT - enet-js is over NINE YEARS old at this point. It's not maintained and it would be very hard to make it in a dead npm libary.

@kuteshiki Sure! What's your name on Discord? I'll add you.

**Comment by kuteshiki** [comment]
𝖐𝖚𝖙𝖊𝖘𝖍𝖎#7592

**Comment by MoanHawks** [comment]
i am not programmer node.js but may I ask some question is that node.js can do copy block of data from source address to a destination address?

**Comment by xXAbood** [comment]
> > @Alexander9673 I'm a node.js / web dev. I'm quite good at node.js and have a discord bot with over 60k (peak) users + a full terminal emulator wroten in HTML + js! (invite.gg/rikka, https://wearetermi.github.io/termi)
> > So, we both could try to copy over the code from GNGS and attempt to do this in node.js.
> > @xXAbood has told me about this converastion.
> > BUT BUT BUT BUT BUT BUT BUT BUT BUT BUT - enet-js is over NINE YEARS old at this point. It's not maintained and it would be very hard to make it in a dead npm libary.
> 
> @kuteshiki Sure! What's your name on Discord? I'll add you.

if u need help u can also dm me on Abood#1337

**Comment by Alexander9673** [comment]
> i am not programmer node.js but may I ask some question is that node.js can do copy a block of data from source address to destination address?

@MoanHawks if you mean by sending specific data from one address to another, yes it is possible. Either through websockets, or http server or other ways.

**Comment by Alexander9673** [comment]
@kuteshiki added :+1: i'm `_alexander#9673`

**Comment by kuteshiki** [comment]
@MoanHawks 
https://github.com/dcodeIO/node-memcpy

**Comment by GrowtopiaNoobs** [comment]
@kuteshiki I don't think there is much problem with age of libraries for enet in other languages, because enet library itself is old and it doesn't receive any important updates

@MoanHawks it isn't really much needed to use memcpy for growtopia server and its usage should be reduced to minimum and also it is possible just to copy both arrays element by element with for loop

**Comment by Alexander9673** [comment]
@GrowtopiaNoobs may i ask if you have a Discord Account? If you do... Can i add you as friend?

**Comment by Alexander9673** [comment]
Having some issues
```js
const enet = require('enet');
const address = { address: '0.0.0.0', port: 17091 };

enet.createServer({
	address: address, /* the address the server host will bind to */
	peers:256, /* allow up to 32 clients and/or outgoing connections */
	channels:2, /* allow up to 2 channels to be used, 0 and 1 */
	down:100, /* assume any amount of incoming bandwidth */
	up:100 /* assume any amount of outgoing bandwidth */

},function(err, host){
  if (err) {
    return console.log(err);
  }

	//setup event handlers..
	host.on("connect",function(peer,data){
    console.log('A Client Connected!')
		//incoming peer connection
		peer.on("message",function(packet,channel){
			console.log("received packet contents:",packet.data());
		});
	});

	//start polling the host for events at 50ms intervals
	host.start(50);
});```

if anyone could help, just reply.
The issue was it's not connecting whatsoever.

**Comment by Alexander9673** [comment]
Tried creating a enet client, it connected. But gt isn't...

**Comment by coldchip** [comment]
GT uses compression and crc32 checksum on enet protocol

**Comment by Alexander9673** [comment]
So what should we do?

**Comment by coldchip** [comment]
The Enet C Library itself has the compression and crc32 checksum function, you need to let nodejs call the function to the native binary, just like how JNI in java works. I recommend you to inplement those inside the nodejs enet packet and the native c enet.

**Comment by kuteshiki** [comment]
Hmm, do you know the function name in the c lib?

**Comment by kuteshiki** [comment]
We can modify the code and call it if needed.

**Comment by coldchip** [comment]
server->checksum = enet_crc32;
	enet_host_compress_with_range_coder(server);

**Comment by DrOreo002** [comment]
gt uses some encryption for its packet. Like what @coldchip has said its crc32 and that compress with range coder I think. You will somehow need to implement that into ur js protocol, maybe copy and paste the checksum code from enet?

**Comment by kuteshiki** [comment]
Thank you so so much! We'll be working on this and will report back!

**Comment by Alexander9673** [comment]
> gt uses some encryption for its packet. Like what @coldchip has said its crc32 and that compress with range coder I think. You will somehow need to implement that into ur js protocol so then gt could understand what you sent, maybe copy and paste the checksum code from enet?

shouldn't gt send like a message to the server when trying to connect or no?

**Comment by coldchip** [comment]
yes, gt sends a packet. but for your case, the server can't understand the packet because of compression.

**Comment by Alexander9673** [comment]
question, can't we just use a udp client or smth? since enet is udp as well or no?

**Comment by Alexander9673** [comment]
Managed to do something without enet. Instead i used `dgram` which is a udp server for nodejs.
It received data like this 
```
<Buffer cf ff 50 0a cc bd 5d f3 83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3a bd ab 28 34 1f f3 83 85>
Received 38 bytes from 127.0.0.1:46930

<Buffer cf ff 52 08 32 47 6e 1d 83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3a bd ab 28 34 1f f3 83 85>
Received 38 bytes from 127.0.0.1:46930

<Buffer cf ff 55 f3 1b 20 ce 9e 83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3a bd ab 28 34 1f f3 83 85>
Received 38 bytes from 127.0.0.1:46930

<Buffer 0f ff ff e0 94 d0 44 ff 00 02 00 00 00 00>
Received 14 bytes from 127.0.0.1:46930

```

when connecting. So i guess it's possible?

**Comment by kuteshiki** [comment]
I've been working on the dgram solution, and here's what I've found out:
all the buffers always start and end with the same thing and length
`
<Buffer cf ff  // this is ALWAYS same
56 f8 8e 2a 9c 40  // always changing
83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3c 57 24 4f d1 0c 8c 72 5f> // same on one user's connection & diffrent on others
`


ALWAYS 38 bytes
client logs out with 
`
<Buffer 0f ff ff e0 94 d0 44 ff 00 02 00 00 00 00>
`

ALWAYS 14 BYTES

FOUND OUT IF USER LEAVE:
`
if(packet.length == 14) return console.log("user left");
`

**Comment by kuteshiki** [comment]
Hmm, I need help. What kind of buffer do I send to the server to make the user join?

**Comment by coldchip** [comment]
You need to use the enet library with the compression enabled, the udp packet is compressed with range coder.

**Comment by coldchip** [comment]
https://github.com/lsalzman/enet/blob/master/compress.c

This is how the buffer is compress in enet

**Comment by DrOreo002** [comment]
> I've been working on the dgram solution, and here's what I've found out:
> 
> buffers start with the same thing (cf ff)
> ` <Buffer cf ff // this is ALWAYS same`
> 
> `56 f8 8e 2a 9c 40 // always changing`
> 
> `83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3c 57 24 4f d1 0c 8c 72 5f> // same on one user's connection & diffrent on others `
> 
> ALWAYS 38 bytes
> 
> client logs out with
> `<Buffer 0f ff ff e0 94 d0 44 ff 00 02 00 00 00 00>`
> 
> ALWAYS 14 BYTES
> 
> FOUND OUT IF USER LEAVE:
> `if(packet.length == 14) return console.log("user left");`

nice job, but that might be because of compression. That first byte maybe an indicator that the packet is compressed and using checksum32. But as coldchip said you will have to implement their compression which wouldn't be so easy. There's no other way apparently, that's also a reason why I made a wrapper in Java (using JNI) and not recreate the whole thing

But well Java connects into native side every time so it wouldn't be a big performance impact on me, I can run a bot that used max 32mb of memory without any leaks with it

**Comment by Bexpg** [comment]
use ur brain 12 yrs old kids cant make growtopia server with node.js if its possible try make minecraft first no one made minecraft with node.js because its impossible

**Comment by DrOreo002** [comment]
> use ur brain 12 yrs old kids cant make growtopia server with node.js if its possible try make minecraft first no one made minecraft with node.js because its impossible

whats wrong with you?

**Comment by Bexpg** [comment]
@DrOreo002  Brainnnnn Dont team with kids

**Comment by DrOreo002** [comment]
> @DrOreo002 Brainnnnn Dont team with kids

I'm trying to help, is that a problem to you?

**Comment by Alexander9673** [comment]
> use ur brain 12 yrs old kids cant make growtopia server with node.js if its possible try make minecraft first no one made minecraft with node.js because its impossible

Yes you can lol. Nodejs is server side. It can communicate with servers. If we can't make it, then how did we manage to get "packets" from Growtopia with a simple udp server setup?

And yes, there is no proper enet for nodejs, and we can't implement one since we don't know that much about C/C++. But we do know what we're doing when it comes to nodejs. And we are trying to find a workaround for the enet encryption.

And i thought, maybe we could implement our own as coldchip sent a link to how enet does it's encryption, we could maybe translate it to nodejs. And for crc32, there are many crc32 modules for nodejs out there.

TL;DR
Nothing is impossible when it comes to coding, you just need to find a workaround.
And don't act like you know everything. Little kid.

**Comment by Alexander9673** [comment]
Another thing, sorry if off-topic, i just don't like seeing people acting like they know everything.
But here's a minecraft server that was made with nodejs.

https://github.com/PrismarineJS/flying-squid/

**Comment by Alexander9673** [comment]
> > I've been working on the dgram solution, and here's what I've found out:
> > buffers start with the same thing (cf ff)
> > ` <Buffer cf ff // this is ALWAYS same`
> > `56 f8 8e 2a 9c 40 // always changing`
> > `83 7b 88 b1 fd f1 df 57 6b 99 9b 36 fc bb 24 3e 8a 64 26 56 d3 3c 57 24 4f d1 0c 8c 72 5f> // same on one user's connection & diffrent on others `
> > ALWAYS 38 bytes
> > client logs out with
> > `<Buffer 0f ff ff e0 94 d0 44 ff 00 02 00 00 00 00>`
> > ALWAYS 14 BYTES
> > FOUND OUT IF USER LEAVE:
> > `if(packet.length == 14) return console.log("user left");`
> 
> nice job, but that might be because of compression. That first byte maybe an indicator that the packet is compressed and using checksum32. But as coldchip said you will have to implement their compression which wouldn't be so easy. There's no other way apparently, that's also a reason why I made a wrapper in Java (using JNI) and not recreate the whole thing
> 
> But well Java connects into native side every time so it wouldn't be a big performance impact on me, I can run a bot that used max 32mb of memory without any leaks with it

So, you're saying that the buffers we got are compressed?

**Comment by DrOreo002** [comment]
> So, you're saying that the buffers we got are compressed?

If you got it from growtopia then yes, if not then maybe it some kind of indicator to specify that packet is from enet or something. But what I know is anything that is on the beginning of a packet could be an indicator of something

**Comment by Alexander9673** [comment]
> > So, you're saying that the buffers we got are compressed?
> 
> If you got it from growtopia then yes, if not then maybe it some kind of indicator to specify that packet is from enet or something. But what I know is anything that is on the beginning of a packet could be an indicator of something

So if those are from gt, we can't decompress it right? And if they are indicators, that's the problem since they're indicators for what?

**Comment by Alexander9673** [comment]
I also found a package that can do range coding,
https://www.npmjs.com/package/compressjs

And then for crc32
https://www.npmjs.com/package/node-crc

**Comment by coldchip** [comment]
this is what is written at the comments of compress.c in enet library

/** 
 @file compress.c
 @brief An adaptive order-2 PPM range coder
*/

**Comment by DrOreo002** [comment]
> So if those are from gt, we can't decompress it right? And if they are indicators, that's the problem since they're indicators for what?

I'm just guessing tho, I do not know really sure. But logically you do not want to decompress something that isn't compressed right?. So an indicator that I'm talking about will be used, like what zlib did they had some few bytes of indicator at the beginning of the compressed file to indicate that the file is compressed with zlib

**Comment by coldchip** [comment]
I think it should be the packet type, sequence id, connection token and the packet data

**Comment by MoanHawks** [comment]
> @DrOreo002 Brainnnnn Dont team with kids

You can't coding c++ even you say you are coming c++ 😳

**Comment by Alexander9673** [comment]
> The Enet C Library itself has the compression and crc32 checksum function, you need to let nodejs call the function to the native binary, just like how JNI in java works. I recommend you to inplement those inside the nodejs enet packet and the native c enet.

While i was trying to find away to get the meaning of those Buffers from the udp4 connection, i tried on enet again. There was a function named `enableCompression()` which was the `host_compress_with_range_coder`. I added it in, still didn't work. Now i think it the problem is on the Growtopia Client's side (STILL NOT SURE, maybe ours) since i created a simple enet client and it detects connections, while when Growtopia tries to connect, nothing happens.

**Comment by coldchip** [comment]
Did you forgot about crc32 checksums?

**Comment by coldchip** [comment]
server->checksum = enet_crc32;

**Comment by Alexander9673** [comment]
Ahh, so gt is not doing anything because of the checksum?

**Comment by coldchip** [comment]
yes

**Comment by Alexander9673** [comment]
okay thank you.

**Comment by Alexander9673** [comment]
but wait, why is it sending data to our `dgram` test? where we didn't use enet, we just created a udp4 socket.

**Comment by coldchip** [comment]
You won't go far using udp4. just use enet on nodejs with compression and crc32 enabled

**Comment by Alexander9673** [comment]
Yeah, problem is we can't find any proper enet for nodejs. Some has compiling issues

**Comment by coldchip** [comment]
I belive there are alot of enet wrapper for nodejs. If you really can't find, the last resort is to make the enet library completely in javascript.

**Comment by DrOreo002** [comment]
^ I wouldn't recommend to remake enet in js tho. That's a bad idea, yes it does possible and clients could also connect but you'll sure not gonna get any of their performance boost. You can do what I did, make a wrapper (ENetJavaWrapper). I literally made a gt proxy with it and it runs so fine that it took only 10mb of my memory

**Comment by coldchip** [comment]
@DrOreo002 actually yea its a bad idea, much slower and javascript lacks threading

**Comment by Alexander9673** [comment]
> ^ I wouldn't recommend to remake enet in js tho. That's a bad idea, yes it does possible and clients could also connect but you'll sure not gonna get any of their performance boost. You can do what I did, make a wrapper (ENetJavaWrapper). I literally made a gt proxy with it and it runs so fine that it took only 10mb of my memory

When making a wrapper for it, do we still need to do some C/C++ Coding or no?

**Comment by coldchip** [comment]
yes you need.

**Comment by Alexander9673** [comment]
Now that's a huge problem, since @kuteshiki codes little in C++, while i don't. Btw, we might make a discord server for this, is it okay if we post the link here to continue discussion there?

**Comment by coldchip** [comment]
okay sure

**Comment by GrowtopiaNoobs** [comment]
@Alexander9673 My discord is Anybody#8639

@Alexander9673 @kuteshiki Please don't try to reimplement enet using raw udp. It contains alot of stuff and would take you alot of time to remake and you would gain nothing.

Also you probably just would need to add two lines of code which enable CRC and range compress. And check out what enet library you are using. There was one version of enet, where compatibility was droped, so make sure you got right one.

**Comment by Alexander9673** [comment]
I'm going to pause this project, since a lot of enet for nodejs is broken. It'll take some time to make a wrapper for it since i'm not a coder in C++. I'll learn it so i could make the wrapper and finish this project.

If you want to talk about it, here's a discord link.
https://discord.gg/3NrVX8s

**Comment by Alexander9673** [comment]
Change of plans. I managed to properly compile [this](https://github.com/linluxiang/node-enet) on an Ubuntu (18.04) VPS. Now will try to compile it on windows and i'll see what happens.

**Comment by Alexander9673** [comment]
> Change of plans. I managed to properly compile [this](https://github.com/linluxiang/node-enet) on an Ubuntu (18.04) VPS. Now will try to compile it on windows and i'll see what happens.

Can't make it compile on Windows, so i had to move the compiled version in the Ubuntu VPS to a Ubuntu VM with the same version. Now, i tested this code
```js
'use strict';
var Buffer = require('buffer').Buffer;
var enet = require('./build/Release/enet');

enet.enet_initialize();

var address = new enet.ENetAddress();
enet.enet_address_set_host(address, '127.0.0.1');
address.port = 17091;

var server = enet.enet_host_create(address, 1024, 2, 0, 0);
enet.enet_host_compress_with_range_coder(server);

var EnetEvent = new enet.ENetEvent();
var enetEvent = enet.enet_host_service(server, EnetEvent, 0);


setInterval(() => {
  while (enetEvent > 0) {
    console.log(checkedEvents)
  }    
}, 10);
```
This code is not the code we're gonna use, we're just testing first, the problem is it won't detect any message from the GT Server. Tell me if i'm doing something wrong.
Btw, the code i'm using is similar to enet itself, so no problem with that.

**Comment by Alexander9673** [comment]
I found out something, it disconnects to the server when i do this:
• Connect to the server using gt.
• Wait until it says Located server connecting...
• Wait a few seconds.
• Click cancel
• Click Quit.

It suddenly disconnects, and it's the only thing we received from gt.
https://hasteb.in/dulavebu.yaml Here's the Enet Event Data

**Comment by Alexander9673** [comment]
Tried to replicate [this](https://github.com/GrowtopiaNoobs/GrowtopiaServer/blob/3ce50c26aa34f3462a385cccff809753c70b3a13/enet%20server%20test/enet%20server%20test.cpp#L2355) but it didn't work, i only replicated the enet part. I tried connecting, i still keep getting event type 0. Anyone knows why?
Code i used:

```js

Save New Duplicate & Edit Just Text
'use strict';
var enet = require('./build/Release/enet');

enet.enet_initialize();

var address = new enet.ENetAddress();
enet.enet_address_set_host(address, '127.0.0.1');

address.port = 17091;
var server = enet.enet_host_create(address, 1024, 10, 0, 0);

enet.enet_host_compress_with_range_coder(server);

var event = new enet.ENetEvent();

var a = function() {
  while(enet.enet_host_service(server, event, 1000) >= 0) {
    var peer = new enet.ENetPeer();
    console.log(event.type)
  }
}

a();```

**Comment by coldchip** [comment]
When the clients connect, the server needs to send this buffer back to the client: 0x01, 0x00, 0x00, 0x00, 0x00

**Comment by coldchip** [comment]
after the client received this buffer, it will show "Logging On..."

**Comment by playingoDEERUX** [comment]
seems like the hello packet with type 1 was successfully sent!

**Comment by Alexander9673** [comment]
> When the client connects, the server needs to send this buffer back to the client: 0x01, 0x00, 0x00, 0x00, 0x00

problem is, i'm not receiving an event which is when the client connected. Or is there one?

**Comment by Alexander9673** [comment]
Here's a video on what's happening.
https://www.youtube.com/watch?v=4K_ZxYw_Jb4

Running the code on a Ubuntu 18.04 Virtual Machine. 
The event type i keep receiving is `0` which means nothing, if the GT Client connected, i should've received `1` right?

**Comment by GrowtopiaNoobs** [comment]
@Alexander9673 Move check for event type into while loop, as it is only valid place to check for that value. Also you are missing crc.

**Comment by MrAugu** [comment]
Serialization of packets will be a pain in ass due to how NodeJS's buffer works, you most likely need to write a sepparate module which handles binary operations with buffers in order to make your life easier.

**Comment by xXAbood** [comment]
@Alexander9673 u might take long with serializing the packets but just keep everything positive since u made it almost halfway

**Comment by GrowtopiaNoobs** [comment]
For those who are interested how to enable CRC in `node-enet`. Add this code to `node-enet.i` after line 18:
```C++
void enableCRC(ENetHost *host) {
    host->checksum = enet_crc32;
}
```
Then you can simply call it from your NodeJS code: `enet.enableCRC(server);`.

**Comment by Alexander9673** [comment]
> For those who are interested how to enable CRC in `node-enet`. Add this code to `node-enet.i` after line 18:
> 
> ```c++
> void enableCRC(ENetHost *host) {
>     host->checksum = enet_crc32;
> }
> ```
> 
> Then you can simply call it from your NodeJS code: `enet.enableCRC(server);`.

As @GrowtopiaNoobs said, to everyone who are intersted, 
here is the `node-enet.i` file https://hasteb.in/fojehupe.shell

After placing replacing the current `node-enet.i` with that code, recompile the module by doing `node-gyp configure build`. Once done, simply run the code.

The function `enetCRC(host)` would be enabled.
Which could be accessed by doing `enet.enetCRC(host)`

**Comment by Alexander9673** [comment]
So... i sent the data, which is `0x01, 0x00, 0x00, 0x00, 0x00`. Now i was wandering, i received a packet. It is in a weird form, 
![Capture](https://user-images.githubusercontent.com/59809478/77539264-0a20be80-6edc-11ea-8d3b-a2138a156bb6.PNG)
looks like that. When i copy the text and pasted it, it said `received packet: 1`

Is it correct? Or am i doing something wrong again?

**Comment by coldchip** [comment]
Print the buffer in hex, the command line doesnt display binary so well

**Comment by Alexander9673** [comment]
Okay, changed it to hex. Got `packet received: 020001310700000000000000000000004139588c4b2f0000b811570300000000
`

What now?

**Comment by MrAugu** [comment]
> Okay, changed it to hex. Got `packet received: 020001310700000000000000000000004139588c4b2f0000b811570300000000 `
> 
> What now?

Is a string which represents bytes (octets) in hexadecimal (base 16).

02 00 01 31 07 00 00 00 00 00 00  00 00 00 00 00 41 39 58 8c 4b 2f 00 00 b8 11 57 03 00 00 00 00

**Comment by Alexander9673** [comment]
Thanks, i alr know about buffers, but thanks.

**Comment by MrAugu** [comment]
You should have a `<packet>.type` property, and based on that you create handlers for each type of packet.

**Comment by Alexander9673** [comment]
@MrAugu so i got the Buffer what kind of deserialize method?

**Comment by MrAugu** [comment]
What's ur code so far?

**Comment by kuteshiki** [comment]
The node.js one or the node-enet.i one?

**Comment by Alexander9673** [comment]
The code is not in final state, just testing some stuff before i rework it. 
```js
'use strict';
var Buffer = require('buffer').Buffer;
var enet = require('./build/Release/enet');

enet.enet_initialize();

var address = new enet.ENetAddress();
enet.enet_address_set_host(address, '0.0.0.0');
address.port = 17091;

var server = enet.enet_host_create(address, 32, 2, 0, 0);
enet.enableCRC(server);
enet.enet_host_compress_with_range_coder(server);

if (server === null) {
  console.log('create server fail');
}

var ioloop = function() {
  var enetEvent = new enet.ENetEvent();
  var error = enet.enet_host_service(server, enetEvent, 0);
  while (error > 0) {
    switch (enetEvent.type) {
      case enet.ENET_EVENT_TYPE_CONNECT:
        console.log('A new client connect');
    var buffer = new Buffer([0x01, 0x00, 0x00, 0x00, 0x00], 'binary');
    var packet = enet.enet_packet_create(buffer.toString(), buffer.length,
    enet.ENET_PACKET_FLAG_RELIABLE);

    enet.enet_peer_send(enetEvent.peer, 0, packet);
    enet.enet_host_flush(server);
        break;
      case enet.ENET_EVENT_TYPE_RECEIVE:
    const buf = Buffer.allocUnsafe(32);
    buf.writeUInt32BE(enetEvent.packet.dataLength, 0);
    buf.write(enet.ConvertToString(enetEvent.packet.data));

    var received = buf.toString('hex');
        console.log(`packet received: ${received}`);
    
    var bytes = received.split(/(?<=^(?:.{2})+)(?!$)/);
    for (var byte of bytes) {
      bytes[bytes.indexOf(byte)] = parseInt(`0x${byte}`, 16)
    }
    
    // would return an array of the value of each byte.
    // console.log(bytes);

    var buffer = new Buffer(bytes, 'binary');

    // this would log the buffer.    
    console.log(buffer);

        enet.enet_packet_destroy(enetEvent.packet);
        break;
    }
    error = enet.enet_host_service(server, enetEvent, 0);
  }
};

setInterval(function() {
  ioloop();
}, 10);

//enet.enet_host_destroy(server);
```

**Comment by Alexander9673** [comment]
Hey guys. Code will be released at Mar 27 9 am GMT +8. Not done yet, i just feel like it's much better to open this to the public.

Also i have a plan on how to save users.
Obv first we need a database, which will support many such as mysql, mongodb, sqlite and etc...

Tho i won't be saving to it every time a user gets updated. Instead, i'll use a cache where it would save player data there and then every xx minutes it would save to the database and remove the player from the cache. If the removed player got updated again, the player's data would be placed in the cache again

**Comment by Alexander9673** [comment]
https://github.com/Alexander9673/node-gtps Here's the code for the private server. It's not done, nor close to done. All i've done so far is the web server, and making the client reach up to `Logging in as ...`

**TAKE NOTE**: Windows support is not yet available due to the fact i'm having trouble compiling it on Windows, Linux works, but you need Node v10.x ...

If anyone could compile it properly on windows, kindly pm me at discord (_alexander#9673)

**Comment by Alexander9673** [comment]
Anyone knows what the enum is for? Specifically `clothTypes`, may i know what is the value of each?

**Comment by Alexander9673** [comment]
Anyways, finished most of the structs. Added them in a folder. No use for now, until i add the code itself that would use it.

Here are the structs: https://github.com/Alexander9673/node-gtps/tree/master/src/structs
Also added password hashing, which is in sha256

**Comment by DrOreo002** [comment]
> Anyone knows what the enum is for? Specifically `clothTypes`, may i know what is the value of each?

I think you can find it on CoreData.txt

**Comment by Alexander9673** [comment]
Also, could you help? Don't really know what enums are since they don't use those in node.
Just need a brief description of what to do.

**Comment by DrOreo002** [comment]
This is what I have when I scrap for data at Growtopia Wiki

[0] = {Hat,Top Hat,Clothes}
[1] = {Shirt,Green Shirt,Clothes}
[2] = {Pants,Jeans,Clothes}
[3] = {Feet,Boots,Clothes}
[4] = {Face,Shades,Clothes}
[5] = {Hand,Pickaxe,Clothes}
[6] = {Back,Fairy Wings,Clothes}
[7] = {Hair,Red Hair,Clothes}
[8] = {Chest,Gold Chain,Clothes}

It was on 2019, its fixed now you can no longer scrap for these data on the wiki. Because they changed on how the data is stored, it didn't store the exact ID now

**Comment by Alexander9673** [comment]
So... I'm planning NOT to translate the code since pointers don't exist in node.js. If possible, could someone give like a step by step, or something similar or directions so i could follow it?

**Comment by MrAugu** [comment]
You don't need pointers, translation is not copying from one spot and pasting it with 1-2 tweaks, it means to reproduce same behavior.

**Comment by Alexander9673** [comment]
Okay then, i'll try to reproduce each function not done.

**Comment by Alexander9673** [comment]
Sorry for asking again but could someone give like a step by step, or something similar or directions so i could follow it?

**Comment by coldchip** [comment]
Maybe Java is easier for you to understand? You can take a look at my repository and copy some functions from there https://github.com/coldchip/growtopiaserver

**Comment by Alexander9673** [comment]
Thanks, i've worked with java before :)

**Comment by Alexander9673** [comment]
Having problems, `packet.data` returns `SwigProxy {}` for me? What is swigproxy?
Also, i can't seem to make your gt server work @coldchip , stuck to located server...

**Comment by coldchip** [comment]
My server port is 10003 thats why it's stuck at locating server.

**Comment by coldchip** [comment]
not sure about packet.data, i abandoned my project long ago. Swig is a JNI generator for Java. It's used to auto create JNI without any hard work.

**Comment by Alexander9673** [comment]
are there any headers i need to add when getting the data from the packet. Cause i'm having a hard time to get the decoded data.

**Comment by Alexander9673** [comment]
After getting packet data, what should we do?

**Comment by coldchip** [comment]
after getting the packet, you need to get the first 4 bytes of the packet. The first 4 bytes is the packet type

Please use little endian instead of big endian when handling with gt packets

**Comment by Alexander9673** [comment]
what would we do with the type?

**Comment by Alexander9673** [comment]
Did some testing, by adding the `GetMessageTypeFromPacket` function into my code, then recompiled it to work on node.js. It does work, and i noticed something. When i login, `GetMessageTypeFromPacket` gives `2`, which i think is correct? And when disconnect it gives `3`

Now, i tried making a function where it gets the data of the packet, but i think it's not really returning the data. Here is the function i did.
```cpp
char* getData(ENetPacket* packet)
{
    return (char*)(packet->data);
}```

Added it in `node-enet.i` then recompiled to work on node. And i noticed something, since it's returning the weird symbol, i turned it into a Buffer. Which then when connecting, return `Buffer <02>` and when disconnected, returned `Buffer <03>`. After receiving those, i turned them into a hex, which would still give the same; `02`, `03`. So now i was wandering, are the `02` and `03` Message Types? Instead of packet data?

**Comment by Alexander9673** [comment]
Also, is it possible to make a function in cpp that returns the data of the packet?

**Comment by coldchip** [comment]
type 2 is normal text packet.
type 3 is binary packet

**Comment by Alexander9673** [comment]
Hey guys, currently i'll make the project private. Will make it public once i finish. Github repo will still be opened, but not updated.

**Comment by GrowtopiaNoobs** [comment]
Maybe this can help you a bit https://github.com/GrowtopiaNoobs/GrowtopiaServer2/tree/master/src/Packets

**Comment by UniC7440** [comment]
May i ask, what part of the code is logging in the game itself until world selection?

**Comment by Alexander9673** [comment]
Hey guys, i might be dropping this project. I managed to get the data by just using GrowtopiaNoob's `GetTextPointerFromPacket`. I know it's been a while, but i havn't really done anything after getting the packet's string form. I don't even have any clue on how or where to continue. Hence, the reason why i have chosen to just drop this project of mine. It was fun through all the problems and i've had. Big thanks to @xXAbood , @MrAugu , @coldchip and especially @GrowtopiaNoobs . Thanks to you guys for helping me when i was in need. Sorry that those help came to a waste, but i had to. Thanks to you guys for the support as well. If anyone is willing to continue this project, kindly open a pull request [here.](https://github.com/Alexander9673/node-gtps) You could also send me a private message on Discord. (_alexander#9673). I'll leave this issue open for a while if anyone have some quick questions, but if none, i'll close it. Thanks again :heart: :v:

**Comment by Alexander9673** [comment]
Also: To those who says "node.js is bad since it's single threaded". Multithreading is actually possible in node.js. You could simple use `child_process` for creating multiple processes or the package `node-webworker-threads`. Which means that Multithreading is possible in node.js.

**Comment by Alexander9673** [comment]
I might be able to cross compile `node-enet`. I'll try this later so if ya'll want to continue it or use it i can give it.

**Comment by LukewarmCat** [comment]
I have a quite a lot of things to do right now, when I get some free time I will work on node-gtps. It's not going to be dropped on my side. Please join [our discord server.](https://discord.gg/zmmqUzj) I am very active on Discord and IRC. I have a account on Freenode by the name "lukewarmcat". Send me a message if you want.

**Comment by Alexander9673** [comment]
Hey guys, i'm back at developing it. So i got some stuff working. Basically i am making a super simple enet wrapper for nodejs which uses crc32 compression and range coder compression. It also uses gtnoob's `packetEnd`, `appendString`, `ch2n`, `createPacket`, `GetTextPointerFromPacket` and `GetMessageTypeFromPacket`. I would maybe add more functions or try to create them myself. This enet wrapper would not be the full enet wrapped for nodejs. As i made it for the purpose of Growtopia. So far i have managed to make the server send messages to the client as seen in this video.

https://www.youtube.com/watch?v=x49Dp8O9xrQ

**Comment by Alexander9673** [comment]
So will now just make a nodejs api for gtnoob's gtps. But will not directly copy -> paste -> modify his code. Instead i'll do them one by one and try to check for some possible optimization. And will also add support for multi threading. Yes multi threading on nodejs. Will also add mysql support.

**The wrapper**: https://github.com/Alexander9673/gtps-wrapper
**The C++ code that the wrapper uses**: https://github.com/Alexander9673/cpp-gtps-wrapper
